plugins/nf-lamin/src/main/nextflow/lamin/LaminObserver.groovy

```diff
@@ -138,118 +123,203 @@ class LaminObserver implements TraceObserver {
+            ]
+        )
+
+        log.info "Run ${run.uid} (https://lamin.ai/${this.instance.getOwner()}/${this.instance.getName()}/transform/${this.transform.uid}/${run.uid})"
+
+        return run
+    // todo: link to project?
     }
 
     protected void finalizeRun() {
-        if (!this.session) throw new IllegalStateException('Session is null')
-        if (!this.transform) throw new IllegalStateException('Transform is null')
-
         WorkflowMetadata wfMetadata = this.session.getWorkflowMetadata()
-        log.debug 'Finalise Run object:\n' +
-            "  run = ln.Run.get(\"${this.run.uid}\")\n" +
-            "  run.finished_at = \"${wfMetadata.complete}\"\n"
-        '  run.environment = ...\n' +
-            '  run.report = ...\n' +
-            '  run.save()\n'
+
+        this.instance.updateRecord(
+            moduleName: 'core',
+            modelName: 'run',
+            uid: this.run.uid,
+            data: [
+                finished_at: wfMetadata.complete,
+                _status_code: wfMetadata.exitStatus
+            ]
+        )
     }
 
-    // TODO: implement tracking an input artifact
-    protected void createInputArtifact(Path path) {
-        // if path is already in workflowInputs, do nothing
-        if (workflowInputs.contains(path)) {
-            return
+    protected Map fetchOrCreateStorage(Path path) {
+        String root = path.getFileSystem().toString()
+        URI uri = path.toUri()
+        String type = uri.getScheme()
+
+        // Search for existing Storage object
+        List<Map> existingStorage = this.instance.getRecords(
+            moduleName: 'core',
+            modelName: 'storage',
+            filter: [
+                and: [
+                    [root: [eq: root]],
+                    [type: [eq: type]]
+                ]
+            ]
+        )
+        log.debug "Found ${existingStorage.size()} existing Storage(s) with root ${root} and type ${type}"
+
+        Map storage = null
+        if (existingStorage) {
+            if (existingStorage.size() > 1) {
+                log.warn "Found multiple Storage objects with root ${root} and type ${type}"
+            }
+            storage = existingStorage[0]
+        } else {
+            // create Storage object
+            storage = this.instance.createRecord(
+                moduleName: 'core',
+                modelName: 'storage',
+                data: [
+                    root: root,
+                    type: type
+                ]
+            )
         }
+        return storage
+    }
 
-        // log.debug "Create Artifact object:\n" +
-        //     "  artifact = ln.Artifact(\n" +
-        //     "    run=run,\n" +
-        //     "    data=\"${path.toUriString()}\",\n" +
-        //     "  )\n"
-        lock.withLock {
-            workflowInputs << path
-        }
+    // TODO: implement tracking an input artifact
+    protected Map createOutputArtifact(Map run, Path localPath, Path destPath) {
```

Comment:
@fredericenard, if we had the "create-artifact" endpoint that calls lamindb, this here would be very simple. Also the storage location management would be taken care off.

Because we'll also enable the "artifact upload" on the UI with it this turns into the highest priority.

- https://github.com/laminlabs/laminhub/issues/2482

Both `nf-lamin` and drag & drop sample sheet submission depend on it.